modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java

```diff
@@ -0,0 +1,391 @@
+     * {@link #handleTriggerKeyUpdate(WatchEvent)} when node receives corresponding events from the metastorage (or if it doesn't receive
+     * this event within a 30 seconds window).
+     */
+    private final Map<UUID, CompletableFuture<Void>> ongoingOperationsById = new ConcurrentHashMap<>();
+
+    /**
+     * Constructor.
+     */
+    public DisasterRecoveryManager(
+            ExecutorService threadPool,
+            MessagingService messagingService,
+            MetaStorageManager metaStorageManager,
+            CatalogManager catalogManager,
+            DistributionZoneManager dzManager,
+            Loza raftManager
+    ) {
+        this.threadPool = threadPool;
+        this.messagingService = messagingService;
+        this.metaStorageManager = metaStorageManager;
+        this.catalogManager = catalogManager;
+        this.dzManager = dzManager;
+        this.raftManager = raftManager;
+
+        watchListener = new WatchListener() {
+            @Override
+            public CompletableFuture<Void> onUpdate(WatchEvent event) {
+                return handleTriggerKeyUpdate(event);
+            }
+
+            @Override
+            public void onError(Throwable e) {
+                // No-op.
+            }
+        };
+
+        messagingService.addMessageHandler(TableMessageGroup.class, this::handleMessage);
+    }
+
+    @Override
+    public CompletableFuture<Void> start() {
+        metaStorageManager.registerExactWatch(RECOVERY_TRIGGER_KEY, watchListener);
+
+        return nullCompletedFuture();
+    }
+
+    @Override
+    public void stop() throws Exception {
+        metaStorageManager.unregisterWatch(watchListener);
+    }
+
+    /**
+     * Updates assignments of the table in a forced manner, allowing for the recovery of raft group with lost majorities. It is achieved via
+     * triggering a new rebalance with {@code force} flag enabled in {@link Assignments} for partitions where it's required. New pending
+     * assignments with {@code force} flag remove old stable nodes from the distribution, and force new Raft configuration via "resetPeers"
+     * so that a new leader could be elected.
+     *
+     * @param zoneId Distribution zone ID.
+     * @param tableId Table ID.
+     * @return Operation future.
+     */
+    public CompletableFuture<Void> resetPartitions(int zoneId, int tableId) {
+        return processNewRequest(new ManualGroupUpdateRequest(UUID.randomUUID(), zoneId, tableId));
+    }
+
+    /**
+     * Returns partition states for all zones' partitions in the cluster. Result is a mapping of {@link TablePartitionId} to the mapping
+     * between a node name and a partition state.
+     *
+     * @param zoneName Zone name.
+     * @return Future with the mapping.
+     */
+    public CompletableFuture<Map<TablePartitionId, Map<String, LocalPartitionState>>> partitionStates(String zoneName) {
+        int latestCatalogVersion = catalogManager.latestCatalogVersion();
```

Comment:
Probably you need to add TODO so as not to forget when implementing catalog compaction.

Reply:
Do we have a JIRA for it?

Reply:
No